### ISSUE TITLE
crate_in_macro_def warning even when decorated with #[allow(...)]

### ISSUE THREAD

**Issue by petekubiak** [rust-lang/rust-clippy#15618]
### Summary

When running `cargo clippy` on a macro referencing `crate`, but with the reference decorated with the corresponding `#[allow()]` I still get a clippy warning for it.

An example is this commit: https://github.com/petekubiak/post-haste/commit/33112ddb1958c83a43343444bba3017d329c2ffe

Running `cargo clippy` shows the warning for line 48 of src/lib.rs (Interestingly it doesn't show for line 77 of the same file)

### Lint Name

crate_in_macro_def

### Reproducer

Checkout this commit https://github.com/petekubiak/post-haste/commit/33112ddb1958c83a43343444bba3017d329c2ffe

run `cargo clippy` in the repo root

(or alternatively `nix develop --command cargo clippy`)

The warning also shows up when the macro is used - running `cargo clippy --example showcase` on the same commit demonstrates this.

### Version

```text
rustc 1.89.0-nightly (ce7e97f73 2025-05-11)
binary: rustc
commit-hash: ce7e97f7371af47e0786f74aa169f6ac9473ff4e
commit-date: 2025-05-11
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.4
```

**Comment by ada4a** [comment]
> Interestingly it doesn't show for line 77 of the same file

That's probably because `#[cfg(target_os = "none")]` is `false`, so the code is `cfg`-d out and as such Clippy doesn't see it -- otherwise it would've probably linted it as well

**Comment by ada4a** [comment]
According to [lint's docs](https://rust-lang.github.io/rust-clippy/master/index.html#crate_in_macro_def):
> Note that if the use of crate is intentional, an allow attribute can be applied to the macro definition, e.g.:
>
> ```rs
> #[allow(clippy::crate_in_macro_def)]
> macro_rules! ok { ... crate::foo ... }
> ```

But I agree that what you tried is the more intuitive option. I'll see if I can enable it as well

**Comment by ada4a** [comment]
It ended up being a bit more complicated than I thought, so I came up with an alternative approach, which I described in Zulip: [#clippy > make &#96;crate_in_macro_def&#96; more intuitive](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/make.20.60crate_in_macro_def.60.20more.20intuitive/with/541318483) -- feel free to comment what you think about it.

**Comment by petekubiak** [comment]
Thanks for investigating this! I must admit I read the documentation, but somehow didn't absorb that the attribute needed to be on the macro itself.

I didn't know about `#[expect()]`.

I'll make the change to my repo and check it works as expected.

**Comment by ada4a** [comment]
> Thanks for investigating this! I must admit I read the documentation, but somehow didn't absorb that the attribute needed to be on the macro itself.

No problem^^ Yeah as I said I can see where you were coming from:)

> I didn't know about `#[expect()]`.

It is pretty cool isn't it:) Though this makes me think – this attribute is generally not very well-known yet, and so there's a risk people will think it's actually some weird thing made just for this lint. I think I'll mention this in the Zulip thread.
